apps/frontend/src/utils/fuzzy-find.ts

```diff
@@ -0,0 +1,226 @@
+// stolen and simplified from: https://github.com/microsoft/vscode/blob/main/src/vs/base/common/fuzzyScorer.ts
+
+const EXACT_MATCH_SCORE = 1 << 18;
+const PREFIX_MATCH_SCORE = 1 << 17;
+const SUBSTRING_MATCH_SCORE = 1 << 16;
+
+const fuzzyScore = (query: string, target: string): number => {
+  // exact match > prefix match > substring match > fuzzy character match > empty
+
+  if (query.trim() === "" || target.trim() === "") return 0; // empty
+
+  // case insensitive
+  query = query.toLowerCase();
+  target = target.toLowerCase();
+
+  // Try exact/prefix/substring matches with space-normalized versions
+  const queryNoSpaces = query.replace(/\s+/g, "");
+  const targetNoSpaces = target.replace(/\s+/g, "");
+
+  if (queryNoSpaces === targetNoSpaces) return EXACT_MATCH_SCORE + query.length; // exact match (ignoring spaces)
+  if (targetNoSpaces.startsWith(queryNoSpaces))
+    return PREFIX_MATCH_SCORE + query.length; // prefix match (ignoring spaces)
+
+  const substringIndex = targetNoSpaces.indexOf(queryNoSpaces);
+  if (substringIndex !== -1) {
+    // Add word boundary bonus for substring matches
+    let wordBoundaryBonus = 0;
+
+    // Check if match is at start of target
+    if (substringIndex === 0) {
+      wordBoundaryBonus = 1000;
+    } else {
+      // Check if match is after a word boundary in original target (with spaces)
+      let actualIndex = 0;
+      let noSpaceIndex = 0;
+      for (let i = 0; i < target.length; i++) {
+        if (target[i] !== " ") {
+          if (noSpaceIndex === substringIndex) {
+            actualIndex = i;
+            break;
+          }
+          noSpaceIndex++;
+        }
+      }
+
+      if (actualIndex > 0 && target[actualIndex - 1] === " ") {
+        wordBoundaryBonus = 500;
```

Comment:
I feel like there should be a better way to do this. If getting an exact substring match like this, we are probably just matching the course number, right (like 16A)? Maybe just do a substring find in the original string and see if the index before the find is a space. For performance